---
title: "6_Fig6e"
author: "Qing Chen"
date: "10/08/2025"
output: 
  html_document:
    number_sections: true
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
      smooth_scroll: false
---

```{r }
rm(list = ls())
getwd()

PATH <- "./"
dir(PATH)
```


# load library
```{r}
library(preprocessCore)
library(scales)
library(ChIPseeker)

library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
library(GenomicRanges)
library(clusterProfiler)

library(org.Mm.eg.db)
# "org.Hs.eg.db"
library(ReactomePA)
# BiocManager::install("ggimage")
# install.packages("Rcpp")
library("ggupset")
library("ggplotify")
library("ggimage")

library(dplyr) # use select
library(tidyr)
library(ggplot2)

library(Hmisc)# in order to use capitalize

## in order to use reduce
library(purrr)
library("pheatmap")
library(GenomicRanges)

```

```{r}
	plot_theme <-  theme_bw() +
		  theme(axis.line = element_line(colour = "black"),
		        panel.grid.major = element_blank(),
		        panel.grid.minor = element_blank(),
		        panel.border = element_blank(),
		        panel.background = element_blank(),
		        # title in the middle
		        plot.title = element_text(hjust = 3),
		        text = element_text(size = 20),axis.text = element_text(size = 20)) +theme(aspect.ratio = 1)

  box_theme <- theme_bw() +
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        # title in the middle
        plot.title = element_text(hjust = 3),
		        text = element_text(size = 20),axis.text = element_text(size = 20)) +theme(aspect.ratio = 1)

```


# 1. load CEBPA clusters

```{r}
Peak_clusters_path <- "./DiffPeakClusters/"
  dir(Peak_clusters_path)
  DiffPeak_clusters <- list()
  Files<-  dir(Peak_clusters_path)[grep(dir(Peak_clusters_path), pattern = "*.bed")]

  for(i in seq(1,length(Files))){
    # i=1
    DiffPeak_clusters[[paste0("C",i)]]<-read.table(paste0(Peak_clusters_path,Files[i]),header = F)

    colnames(DiffPeak_clusters[[paste0("C",i)]]) <- c("chr","start","end","idx")
  }
  lapply(DiffPeak_clusters, dim)
```


```{r}
head(DiffPeak_clusters[["C1"]])
```
#2. load two big cluster of DEGs 



```{r}


Cluster1_down <- read.table("./Cluster1_down_genes.txt")$V1
length(Cluster1_down)
Cluster2_up <- read.table("./Cluster2_Up_genes.txt")$V1
length(Cluster2_up)
```
```{r}
TPM<- read.csv("./TPM.csv")
TPM <- select(TPM,-X)

names(TPM)
```


```{r}
Ctrl_mean <- apply(TPM[,2:4],1,mean)
RA4d_mean <- apply(TPM[,11:13],1,mean)
length(Ctrl_mean)
length(RA4d_mean)

up <- TPM[ RA4d_mean>Ctrl_mean,]$Gene
down <- TPM[ RA4d_mean<Ctrl_mean,]$Gene
length(up)
length(down)
Cluster2_up <- up
Cluster1_down <- down
```
### Promoter: DEG TSS+/- 1kb

#### load gene region and extract TSS based on strand 



```{r}
hg19_gtf_reduce<- read.table("./hg19_2015_reduced.bed",sep ="\t",header = T)

dim(hg19_gtf_reduce)
head(hg19_gtf_reduce)
```



```{r}
# get TSS 

hg19_gtf_reduce[["TSS"]] <- ifelse(hg19_gtf_reduce$strand == "-",hg19_gtf_reduce$end,hg19_gtf_reduce$start)
tail(hg19_gtf_reduce)

hg19_TSS<- hg19_gtf_reduce[,c("gene_name","TSS","strand","seqnames")]

head(hg19_TSS)

dim(hg19_TSS)
```

#### extend TSS +/-1kb ( no negative values )
```{r}
d = 1000

hg19_TSS[["TSS_Up1k"]] <- ifelse(hg19_TSS$strand == "-",hg19_TSS$TSS + d,hg19_TSS$TSS-d)

hg19_TSS[["TSS_Down1k"]] <- ifelse(hg19_TSS$strand == "-",hg19_TSS$TSS - d,hg19_TSS$TSS+d)

head(hg19_TSS)

hg19_TSS[which(hg19_TSS$TSS_Up1k <0),"TSS_Up1k"] <- 0
hg19_TSS[which(hg19_TSS$TSS_Down1k <0),"TSS_Down1k"] <- 0

Promoters <- hg19_TSS 
```

### add TSS +/-1k to DEG clusters


```{r}
Cluster2_up_TSS1k<-Promoters[which(Promoters$gene_name %in% Cluster2_up ),c("seqnames","TSS_Up1k","TSS_Down1k","gene_name")]
dim(Cluster2_up_TSS1k) ## more than the gene #

length(Cluster2_up)

chromosomes <-  c(paste0("chr",1:22),"chrX","chrY" )

Cluster2_up_TSS1k <- Cluster2_up_TSS1k[which(Cluster2_up_TSS1k$seqnames %in% chromosomes),]

cat("missed genes: \n")
dim(Cluster2_up_TSS1k)[1]- length(Cluster2_up)  # still miss some genes



Cluster2_up_TSS1k["start"]<-apply(Cluster2_up_TSS1k[,c("TSS_Up1k","TSS_Down1k")],1,min)
Cluster2_up_TSS1k["end"]<-apply(Cluster2_up_TSS1k[,c("TSS_Up1k","TSS_Down1k")],1,max)

Cluster2_up_TSS1k <- Cluster2_up_TSS1k[,c("seqnames","start","end","gene_name")]
head(Cluster2_up_TSS1k)

```

# 3. up DEG associate CEBPA peaks



```{r}
# two clusters of DEGs
names(Cluster1_down_TSS1k)<- c("chr","start","end","Gene")
names(Cluster2_up_TSS1k)<- c("chr","start","end","Gene")
head(Cluster2_up_TSS1k)
```

## CEBPA TPM (signal) and clusters

```{r}
# all peaks
Peak <- read.csv("./CEBPA_merged_pooled_Peaks.bed",sep= "\t", header = F)
names(Peak) <- c("chr","start","end","idx")
dim(Peak)
```


```{r}
CEBPA <- read.csv("./DiffPeak/TPM_RIP_norm.csv")
dim(CEBPA)
head(CEBPA)
```

```{r}
CEBPA_TPM <- cbind(Peak,CEBPA)
head(CEBPA_TPM)
```






```{r}
df_list <- list()

for (C in names(DiffPeak_clusters)){
  
  # C = names(DiffPeak_clusters)[1]
  cat(C,"\n")
  
  ov<- findOverlaps(GRanges(DiffPeak_clusters[[C]]),GRanges(Cluster2_up_TSS1k))
  
  OV_df <- data.frame(DiffPeak_clusters[[C]][ov@from,],Cluster2_up_TSS1k[ov@to,])
  
  dim(OV_df)
  OV_df<- merge(x= OV_df,y= CEBPA_TPM,by="idx",all.x = T)
  df_list[[C]] <- OV_df
  
}
```
```{r}
Up_DEG_CEBPA <- Reduce(full_join,df_list)
dim(Up_DEG_CEBPA)

```

```{r}
head(Up_DEG_CEBPA)
```



```{r}
Up_DEG_CEBPA_unique <- Up_DEG_CEBPA[!duplicated(Up_DEG_CEBPA$idx),]
dim(Up_DEG_CEBPA_unique)

names(Up_DEG_CEBPA_unique)
```


```{r}
CEBPA_df <- Up_DEG_CEBPA_unique[c(1,12:20)]
rownames(CEBPA_df) <- seq(1,dim(CEBPA_df)[1])
```
#------------------------
## ( 0 ): CEBPA peaks that overlap with  Up DEGs Tss1k (re-run code to get it)

```{r}
CEBPA_df
```

## (1) input DEGs Tss1k that overlap with CEBPA peaks

```{r}
Up_DEG_CEBPA_unique <- read.csv("./CEBPA_ov_Up_DEGs_TSS1k.csv")
dim(Up_DEG_CEBPA_unique)
```

## (2) Those DEG TPM
```{r}
RNA_TPM <- read.csv("./TPM.csv")
RNA_TPM<- select(RNA_TPM,-X)
head(RNA_TPM)
```

```{r}
RNA_df <-  RNA_TPM[which(RNA_TPM$Gene %in% Up_DEG_CEBPA_unique$Gene ),]
dim(RNA_df)
rownames(RNA_df) <- seq(1,dim(RNA_df)[1])
```

## (3) Up DEG associated ATAC



```{r}
## all Gene data
ATAC <- read.csv("/Users/kai/Library/CloudStorage/GoogleDrive-qchen2019@gwmail.gwu.edu/My Drive/0_Personal/00_Project/PHA/00_StartOver_2022/ATACseq/HL60/DiffPeak/TPM_RIP_norm.csv")

dim(ATAC)
```


```{r}
peak <- read.csv("/Users/kai/Library/CloudStorage/GoogleDrive-qchen2019@gwmail.gwu.edu/My Drive/0_Personal/00_Project/PHA/00_StartOver_2022/ATACseq/HL60/HL60_merged_pooled_Peaks.bed",sep = "\t",header = F)
dim(peak)
names(peak) <- c("chr","start","end","idx")

```


```{r}

ATAC_TPM <- cbind(peak,ATAC)
head(ATAC_TPM)
```

```{r}


  ov<- findOverlaps(GRanges(ATAC_TPM),GRanges(Up_DEG_CEBPA_unique))
  
  OV_df <- data.frame(ATAC_TPM[ov@from,],Up_DEG_CEBPA_unique[ov@to,])
  
  dim(OV_df)

```

```{r}
ATAC_df <- ATAC_TPM[ov@from,4:dim(ATAC_TPM)[2]]
rownames(ATAC_df) <- seq(1,dim(ATAC_df)[1])
```


## 5. plot heatmap for each individual data set 

```{r}
dim(RNA_df)

dim(ATAC_df)
dim(CEBPA_df)

```


## mean value, then z-score
### get mean
```{r}
head(CEBPA_df)
```


```{r}
CEBPA <- CEBPA_df %>%
  pivot_longer(cols = -idx, 
               names_to = c("remove","Condition", "Replicate"), 
               names_sep = "_") 
CEBPA <- select(CEBPA, -remove)
CEBPA$type <- "CEBPA"
head(CEBPA)
```


```{r}
CEBPA_mean <- CEBPA %>%
  group_by(idx,type, Condition) %>%
  summarise(Mean_Expression = mean(value)) %>%
  pivot_wider(names_from = Condition, values_from = Mean_Expression)

head(CEBPA_mean)
```



```{r}
names(CEBPA_mean) <- c("idx","type","RA_0h","RA_4d","RA_4h")
head(CEBPA_mean)
```

```{r}
head(RNA_df)
RNA_df$type <- "RNA"
```

```{r}

RNA_mean <- RNA_df %>%
  pivot_longer(cols = -c(Gene,type), 
               names_to = c("Condition", "Replicate"), 
               names_sep = "_") %>%
  group_by(Gene,type, Condition) %>%
  summarise(Mean_Expression = mean(value)) %>%
  pivot_wider(names_from = Condition, values_from = Mean_Expression)

head(RNA_mean)
```
```{r}
colnames(RNA_mean) <- c("Gene","type","RA_0h","RA_2d","RA_4d","RA_4h")
```


```{r}
head(ATAC_df)
```

```{r}
ATAC <- ATAC_df %>%
  pivot_longer(cols = -c(idx), 
               names_to = c("remove","Condition", "Replicate"), 
               names_sep = "_") 
ATAC <- select(ATAC, -remove)


ATAC$type <- "ATAC"
head(ATAC)
```


```{r}
ATAC_mean <- ATAC %>%
  group_by(Condition,idx,type) %>%
  summarise(Mean_Expression = mean(value)) %>%
  pivot_wider(names_from = Condition, values_from = Mean_Expression)

head(ATAC_mean)
```


```{r}
names(ATAC_mean) <- c("idx","type","RA_2d","RA_4d","RA_4h","RA_0h")
length(unique(ATAC_mean$idx))
```

## plot
```{r}
pdf("./up_DEG_asso_data_heatmaps_mean_Z-score_hierarchy.pdf", width = 8, height = 4)


### RNA 


RNA_mean_rm2d <- RNA_mean[, -grep("2d|type", names(RNA_mean))]
names(RNA_mean_rm2d)
RNA_mean_rm2d<- RNA_mean_rm2d[,c(1,2,4,3)]
df_temp <- RNA_mean_rm2d
df_temp <- df_temp[which(apply(df_temp[,c(-1)],1,var) != 0),]
 
	mtx <- as.matrix(df_temp[,c(-1)])
  rownames(mtx) <- df_temp$Gene
  head(mtx)
  
  mtx_colnames <-  colnames(df_temp[,c(-1)])
  mtx_colnames

  # fun1 <- function(x) {return (strsplit(x,split = "_")[[1]][2])}
  # Conditions <- (unlist(lapply(mtx_colnames,fun1)))
Conditions <-mtx_colnames
  Sample_info <- data.frame(Conditions  = factor(Conditions,levels = c("RA_0h","RA_4h","RA_4d")))

  rownames(Sample_info) <- mtx_colnames

  pheatmap(mtx ,cluster_rows=F,cluster_cols = F,
           # show the gene name
           show_rownames = F,
           scale = "row" ,
           # cutree_rows=5,
           angle_col='90',
           # annotation_row = gene_anno , 
           annotation_col = Sample_info,
           fontsize_col=10,
           main = paste("RNA mean \n"))

### ATAC


ATAC_mean_rm2d <- ATAC_mean[, -grep("2d|type", names(ATAC_mean))]
names(ATAC_mean_rm2d)

ATAC_mean_rm2d<- ATAC_mean_rm2d[,c(1,4,3,2)]
names(ATAC_mean_rm2d)

df_temp <- ATAC_mean_rm2d

df_temp <- df_temp[which(apply(df_temp[,c(-1)],1,var) != 0),]
 
	mtx <- as.matrix(df_temp[,c(-1)])
  rownames(mtx) <- df_temp$idx
  head(mtx)
  
  mtx_colnames <-  colnames(df_temp[,c(-1)])
  mtx_colnames

Conditions <-mtx_colnames
  Sample_info <- data.frame(Conditions  = factor(Conditions,levels = c("RA_0h","RA_4h","RA_4d")))

  ##  IMPORTANT ******* here must be consistent ********!!!!!!!
  ## colnames(mtx) should be matched with rownames(df)
  rownames(Sample_info) <- mtx_colnames

  pheatmap(mtx ,cluster_rows=F,cluster_cols = F,
           # show the gene name
           show_rownames = F,
           scale = "row",
           # cutree_rows=5,
           angle_col='90',
           # annotation_row = gene_anno , 
           annotation_col = Sample_info,
           fontsize_col=10,
           main = paste("ATAC mean \n"))

    
    


### CEBPA



df_temp <- CEBPA_mean
df_temp<- df_temp[,c(1,3,5,4)]
names(df_temp)
df_temp <- df_temp[which(apply(df_temp[,c(-1)],1,var) != 0),]
 
	mtx <- as.matrix(df_temp[,c(-1)])
  rownames(mtx) <- df_temp$idx
  head(mtx)
  
  mtx_colnames <-  colnames(df_temp[,c(-1)])
  mtx_colnames

Conditions <-mtx_colnames
  Sample_info <- data.frame(Conditions  = factor(Conditions,levels = c("RA_0h","RA_4h","RA_4d")))


  ##  IMPORTANT ******* here must be consistent ********!!!!!!!
  ## colnames(mtx) should be matched with rownames(df)
  rownames(Sample_info) <- mtx_colnames

  pheatmap(mtx ,cluster_rows=F,cluster_cols = F,
           # show the idx name
           show_rownames = F,
           scale = "row",
           # cutree_rows=5,
           angle_col='90',
           # annotation_row = idx_anno , 
           annotation_col = Sample_info,
           fontsize_col=10,
           main = paste(" CEBPA mean\n"))

    


dev.off()
```
