{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5d09b77c",
   "metadata": {},
   "source": [
    "## goal\n",
    "\n",
    "ChrInt_boxplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "d4d7aa00",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import hicstraw\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "from matplotlib import backends\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import matplotlib.pyplot as plt \n",
    "import seaborn as sns\n",
    "import math\n",
    "from matplotlib import rcParams\n",
    "import os\n",
    "import re\n",
    "import scipy\n",
    "\n",
    "from statannot import add_stat_annotation\n",
    "from statannotations.Annotator import Annotator"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd7583f4",
   "metadata": {},
   "source": [
    "## take a look at  hic data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "f3e3b19f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hic = hicstraw.HiCFile(\"https://www.encodeproject.org/files/ENCFF718AWL/@@download/ENCFF718AWL.hic\")\n",
    "hic_file = \"/home/qchen/PhD/Neu/OCEANC/Neu_H/hic_results/H_hicpro2juicebox/HL60_Ctrl.allValidPairs.hic\"\n",
    "hic_test = hicstraw.HiCFile(hic_file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "b69a1452",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/HiC-Pro_3.0.0/annotation/chrom_hg19.sizes\n",
      "[2500000, 1000000, 500000, 250000, 100000, 50000, 25000, 10000, 5000]\n"
     ]
    }
   ],
   "source": [
    "print(hic_test.getGenomeID())\n",
    "print(hic_test.getResolutions())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0202bf6d",
   "metadata": {},
   "source": [
    "#### "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "4f7a2e47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ALL': 3095693,\n",
       " '1': 249250621,\n",
       " '2': 243199373,\n",
       " '3': 198022430,\n",
       " '4': 191154276,\n",
       " '5': 180915260,\n",
       " '6': 171115067,\n",
       " '7': 159138663,\n",
       " '8': 146364022,\n",
       " '9': 141213431,\n",
       " '10': 135534747,\n",
       " '11': 135006516,\n",
       " '12': 133851895,\n",
       " '13': 115169878,\n",
       " '14': 107349540,\n",
       " '15': 102531392,\n",
       " '16': 90354753,\n",
       " '17': 81195210,\n",
       " '18': 78077248,\n",
       " '19': 59128983,\n",
       " '20': 63025520,\n",
       " '21': 48129895,\n",
       " '22': 51304566,\n",
       " 'X': 155270560,\n",
       " 'Y': 59373566,\n",
       " 'M': 16571}"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom_length = {}\n",
    "for chrom in hic_test.getChromosomes():\n",
    "    # print(chrom.name, chrom.length)\n",
    "    chrom_length[chrom.name] = chrom.length\n",
    "chrom_length \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d39eed9",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "del chrom_length[\"ALL\"]\n",
    "\n",
    "del chrom_length[\"M\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "8f1c6f1f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', 'X', 'Y'])"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chrom_length.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "349831bc-d998-4544-bab4-020f76282aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# extract All ChrInt of hub anchor associated with a gene (several genes)\n",
    "\n",
    "\n",
    "def All_ChrInt_in_a_hub(hubs_anchor, hic_getMatrixZoomData_dict, chrom_length, norm, resolution):   \n",
    "\n",
    "    contact = {}\n",
    "\n",
    "    ## step 1: go through every anchors and locate the bin \n",
    "\n",
    "    for i in range(hubs_anchor.shape[0]):\n",
    "        # two anchor in the same chr\n",
    "        Chr = hubs_anchor.iloc[i,0]\n",
    "        \n",
    "        Start1 =  hubs_anchor.iloc[i,1]\n",
    "        End1 = hubs_anchor.iloc[i,2]\n",
    "        \n",
    "        Start2 =  hubs_anchor.iloc[i,4]\n",
    "        End2 = hubs_anchor.iloc[i,5]\n",
    "    ## step 2: get the small matrix of the anchor\n",
    "        \n",
    "        # matrix_chr = hic.getMatrixZoomData( Chr[3:], Chr[3:], \"oe\", norm, \"BP\", resolution)\n",
    "        matrix_chr = hic_getMatrixZoomData_dict[Chr]\n",
    "        \n",
    "        Start1_bin = int(Start1/resolution)*resolution\n",
    "        End1_bin = int(End1/resolution)*resolution\n",
    "        \n",
    "        Start2_bin = int(Start2/resolution)*resolution\n",
    "        End2_bin = int(End2/resolution)*resolution\n",
    "        \n",
    "        numpy_matrix_anchor = matrix_chr.getRecordsAsMatrix(Start1_bin, End1_bin ,Start2_bin, End2_bin)\n",
    "        # numpy_matrix_anchor : does not necessarily symmetric. only if two anchors are the same length\n",
    "        \n",
    "        print(numpy_matrix_anchor.shape)\n",
    "        \n",
    "\n",
    "    ## step 3: turn the symmetric matrix to arrary & save the data to the dict\n",
    "        Gene = hubs_anchor.iloc[i,6]\n",
    "   \n",
    "        if numpy_matrix_anchor.shape[0] == numpy_matrix_anchor.shape[1]:\n",
    "            contact[Gene] = numpy_matrix_anchor[np.triu_indices(numpy_matrix_anchor.shape[0])]\n",
    "            print(Gene + \" has \" + str(len(numpy_matrix_anchor[np.triu_indices(numpy_matrix_anchor.shape[0])])))\n",
    "        ##---- rectangle ----\n",
    "        else: \n",
    "            contact[Gene] = numpy_matrix_anchor.flatten()\n",
    "            print(Gene + \" has \" + str(len(numpy_matrix_anchor.flatten())))\n",
    "        \n",
    "\n",
    "    print(\"Number of genes we are ploting: \"+str(len(contact)))\n",
    "\n",
    "\n",
    "    return  contact"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "309a01b7",
   "metadata": {},
   "source": [
    "## run on the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "c0d7e887-b616-49a6-a5d6-250d245531c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HL60_2d.allValidPairs.hic\n",
      "HL60_4d.allValidPairs.hic\n",
      "HL60_4h.allValidPairs.hic\n",
      "HL60_Ctrl.allValidPairs.hic\n"
     ]
    }
   ],
   "source": [
    "! ls /home/qchen/PhD/Neu/OCEANC/Neu_H/hic_results/H_hicpro2juicebox | grep -v \"rep\" | grep \".allValidPairs.hic\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "cb3955f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here DO NOT add \\ in My Drive\n",
    "\n",
    "Gene_asso_hub_file = '/home/qchen/PhD/Neu/Gene_asso_region_ChrInt_boxplot/Gene_asso_region_ForPlot_add.csv'\n",
    "# Gene_asso_hub_file = '/home/qchen/PhD/ExhaustedT/hichub_asso_CTCF_ATAC/'\n",
    "\n",
    "hic_path = '/home/qchen/PhD/Neu/OCEANC/Neu_H/hic_results/H_hicpro2juicebox/'\n",
    "# ! ls /home/qchen/PhD/Neu/OCEANC/Neu_H/hic_results/H_hicpro2juicebox | grep -v \"rep\" | grep \".allValidPairs.hic\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "dc2da67c-42f3-44b5-a0f5-91dfef9a3aa4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/qchen/PhD/Neu/Gene_asso_region_ChrInt_boxplot/Gene_asso_region_ForPlot_rm_4d.csv\n"
     ]
    }
   ],
   "source": [
    "! ls $Gene_asso_hub_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19eb0734",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "norm = \"KR\"\n",
    "resolution = 10**4\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4696554-d091-4634-832f-849769465f44",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "hubs_anchor = pd.read_csv(Gene_asso_hub_file,header = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7e44fde-348a-48a6-9fce-4aa182624473",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "32684602",
   "metadata": {},
   "source": [
    "## calculate all ChrInt in 4 conditions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "63a3b0c7-fb2e-4862-97a8-6ebc2e247bf4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HL60_2d.allValidPairs.hic\n",
      "HL60_4d.allValidPairs.hic\n",
      "HL60_4h.allValidPairs.hic\n",
      "HL60_Ctrl.allValidPairs.hic\n"
     ]
    }
   ],
   "source": [
    "# ls /home/qchen/PhD/ExhaustedT/Hic/results/hic_hicpro2juicebox |grep -v \"dist\"|grep \"m.hic\"\n",
    "\n",
    "! ls /home/qchen/PhD/Neu/OCEANC/Neu_H/hic_results/H_hicpro2juicebox | grep -v \"rep\" | grep \".allValidPairs.hic\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "c2a3d28a-08a3-4877-b2f9-ece535b278dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HL60_4d.allValidPairs.hic',\n",
       " 'HL60_4h.allValidPairs.hic',\n",
       " 'HL60_Ctrl_rep2.allValidPairs.hic',\n",
       " 'HL60_2d_rep1.allValidPairs.hic',\n",
       " 'HL60_4h_rep1.allValidPairs.hic',\n",
       " 'HL60_4h_rep2.allValidPairs.hic',\n",
       " 'HL60_2d.allValidPairs.hic',\n",
       " 'HL60_4d_rep1.allValidPairs.hic',\n",
       " 'HL60_Ctrl_rep1.allValidPairs.hic',\n",
       " 'HL60_2d_rep2.allValidPairs.hic',\n",
       " 'HL60_4d_rep2.allValidPairs.hic',\n",
       " 'HL60_Ctrl.allValidPairs.hic']"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Hic_files = [f for f in os.listdir(hic_path) if re.match(r'.*.allValidPairs.hic', f)]\n",
    "Hic_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "48ddc199-f8cb-405f-a95d-68ea0e2f1c30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HL60_Ctrl.allValidPairs.hic',\n",
       " 'HL60_4h.allValidPairs.hic',\n",
       " 'HL60_2d.allValidPairs.hic',\n",
       " 'HL60_4d.allValidPairs.hic']"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# myorder = [3,0,2,1,4]\n",
    "myorder = [11,1,6,0]\n",
    "# myorder = [11,1,6]\n",
    "Hic_files = [Hic_files[i] for i in myorder]\n",
    "Hic_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "83d97032-94aa-4efc-8566-a401819fa044",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "HL60_Ctrl.allValidPairs.hic\n",
      "1\n",
      "HL60_4h.allValidPairs.hic\n",
      "2\n",
      "HL60_2d.allValidPairs.hic\n",
      "3\n",
      "HL60_4d.allValidPairs.hic\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(Hic_files)):\n",
    "    print(i)\n",
    "    print(Hic_files[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f316994c-b319-4f0d-9bc1-417350b9369b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HL60_Ctrl.allValidPairs.hic\n",
      "HL60_Ctrl\n",
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chrX\n",
      "chrY\n",
      "(10, 10)\n",
      "CYBB has 55\n",
      "(24, 24)\n",
      "CCL2_1 has 300\n",
      "(8, 10)\n",
      "CCL2_2 has 80\n",
      "(14, 7)\n",
      "CCL2_3 has 98\n",
      "(8, 8)\n",
      "CCL2_4 has 36\n",
      "(10, 10)\n",
      "CEBPE_1 has 55\n",
      "(22, 22)\n",
      "CEBPE_2 has 253\n",
      "(22, 22)\n",
      "ALOX5AP has 253\n",
      "(50, 50)\n",
      "CEBPB_1 has 1275\n",
      "(28, 28)\n",
      "CEBPB_2 has 406\n",
      "(7, 36)\n",
      "CEBPB_3 has 252\n",
      "(10, 10)\n",
      "IRF5 has 55\n",
      "(9, 9)\n",
      "CEBPA has 45\n",
      "Number of genes we are ploting: 13\n",
      "HL60_4h.allValidPairs.hic\n",
      "HL60_4h\n",
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chrX\n",
      "chrY\n",
      "(10, 10)\n",
      "CYBB has 55\n",
      "(24, 24)\n",
      "CCL2_1 has 300\n",
      "(8, 10)\n",
      "CCL2_2 has 80\n",
      "(14, 7)\n",
      "CCL2_3 has 98\n",
      "(8, 8)\n",
      "CCL2_4 has 36\n",
      "(10, 10)\n",
      "CEBPE_1 has 55\n",
      "(22, 22)\n",
      "CEBPE_2 has 253\n",
      "(22, 22)\n",
      "ALOX5AP has 253\n",
      "(50, 50)\n",
      "CEBPB_1 has 1275\n",
      "(28, 28)\n",
      "CEBPB_2 has 406\n",
      "(7, 36)\n",
      "CEBPB_3 has 252\n",
      "(10, 10)\n",
      "IRF5 has 55\n",
      "(9, 9)\n",
      "CEBPA has 45\n",
      "Number of genes we are ploting: 13\n",
      "HL60_2d.allValidPairs.hic\n",
      "HL60_2d\n",
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chrX\n",
      "chrY\n",
      "(10, 10)\n",
      "CYBB has 55\n",
      "(24, 24)\n",
      "CCL2_1 has 300\n",
      "(8, 10)\n",
      "CCL2_2 has 80\n",
      "(14, 7)\n",
      "CCL2_3 has 98\n",
      "(8, 8)\n",
      "CCL2_4 has 36\n",
      "(10, 10)\n",
      "CEBPE_1 has 55\n",
      "(22, 22)\n",
      "CEBPE_2 has 253\n",
      "(22, 22)\n",
      "ALOX5AP has 253\n",
      "(50, 50)\n",
      "CEBPB_1 has 1275\n",
      "(28, 28)\n",
      "CEBPB_2 has 406\n",
      "(7, 36)\n",
      "CEBPB_3 has 252\n",
      "(10, 10)\n",
      "IRF5 has 55\n",
      "(9, 9)\n",
      "CEBPA has 45\n",
      "Number of genes we are ploting: 13\n",
      "HL60_4d.allValidPairs.hic\n",
      "HL60_4d\n",
      "chr1\n",
      "chr2\n",
      "chr3\n",
      "chr4\n",
      "chr5\n",
      "chr6\n",
      "chr7\n",
      "chr8\n",
      "chr9\n",
      "chr10\n",
      "chr11\n",
      "chr12\n",
      "chr13\n",
      "chr14\n",
      "chr15\n",
      "chr16\n",
      "chr17\n",
      "chr18\n",
      "chr19\n",
      "chr20\n",
      "chr21\n",
      "chr22\n",
      "chrX\n",
      "chrY\n",
      "(10, 10)\n",
      "CYBB has 55\n",
      "(24, 24)\n",
      "CCL2_1 has 300\n",
      "(8, 10)\n",
      "CCL2_2 has 80\n",
      "(14, 7)\n",
      "CCL2_3 has 98\n",
      "(8, 8)\n",
      "CCL2_4 has 36\n",
      "(10, 10)\n",
      "CEBPE_1 has 55\n",
      "(22, 22)\n",
      "CEBPE_2 has 253\n",
      "(22, 22)\n",
      "ALOX5AP has 253\n",
      "(50, 50)\n",
      "CEBPB_1 has 1275\n",
      "(28, 28)\n",
      "CEBPB_2 has 406\n",
      "(7, 36)\n",
      "CEBPB_3 has 252\n",
      "(10, 10)\n",
      "IRF5 has 55\n",
      "(9, 9)\n",
      "CEBPA has 45\n",
      "Number of genes we are ploting: 13\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "Hic_contact_list = []\n",
    "\n",
    "    \n",
    "for H in Hic_files:\n",
    "    print(H)\n",
    "    # 1. for each hic file, calculated the gene asso hub ChrInt \n",
    "    print(H.split('.')[0])\n",
    "    \n",
    "    #  ---> load corresponding hic data\n",
    "    # hic = hicstraw.HiCFile(hic_path + H )\n",
    "    hic_getMatrixZoomData_dict = hic_getMatrixZoomData(hic_path + H)\n",
    "\n",
    "    \n",
    "    Hic_contact_list.append(All_ChrInt_in_a_hub(hubs_anchor, hic_getMatrixZoomData_dict, chrom_length, norm, resolution))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "8b52abf8-b4a4-42f9-afb4-2ad99b06a20d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Hic_contact_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a68ac8f-010c-4fee-98aa-728362b76ddd",
   "metadata": {},
   "source": [
    "\n",
    "# plot\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5b9fe54-f0c5-4718-ad7c-84fe2a54b13b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CYBB\n",
      "0h vs 4h RanksumsResult(statistic=-0.5529784118452927, pvalue=0.5802781928282779)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.6828478155075666, pvalue=0.09240455591213687)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.6828478155075666, pvalue=0.09240455591213687)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=5.996e-01 stat=6.170e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=8.112e-03 stat=4.120e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=6.780e+02\n",
      "1.1172130534195424 0.34299679463060395\n",
      "CCL2_1\n",
      "0h vs 4h RanksumsResult(statistic=-1.6631439865907918, pvalue=0.09628361679032632)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.8468387344725274, pvalue=0.06477051618298572)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.8468387344725274, pvalue=0.06477051618298572)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.159e-01 stat=1.946e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=7.611e-02 stat=1.921e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=2.165e+04\n",
      "1.1025100178246179 0.3470383549751718\n",
      "CCL2_2\n",
      "0h vs 4h RanksumsResult(statistic=-1.600517918892398, pvalue=0.10948373500486415)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.7438478817782843, pvalue=0.08118561864708883)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.7438478817782843, pvalue=0.08118561864708883)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=3.028e-01 stat=1.278e+03\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.965e-01 stat=1.236e+03\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=1.512e+03\n",
      "1.091139428280968 0.3530059254096827\n",
      "CCL2_3\n",
      "0h vs 4h RanksumsResult(statistic=-1.984529923451954, pvalue=0.04719679566000956)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.3184028108211903, pvalue=0.1873688403617032)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.3184028108211903, pvalue=0.1873688403617032)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=4.409e-02 stat=1.737e+03\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=5.158e-01 stat=2.040e+03\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=2.228e+03\n",
      "0.7248125012854566 0.5376518343694711\n",
      "CCL2_4\n",
      "0h vs 4h RanksumsResult(statistic=-1.2613764396790406, pvalue=0.2071732526765151)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.5204091013988434, pvalue=0.12840818844495236)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.5204091013988434, pvalue=0.12840818844495236)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=2.820e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=2.476e-01 stat=2.220e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=6.261e-01 stat=2.520e+02\n",
      "0.7545306308701552 0.5214929058153557\n",
      "CEBPE_1\n",
      "0h vs 4h RanksumsResult(statistic=-0.03287979746107146, pvalue=0.9737704433748545)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.001339286314449, pvalue=0.3166628057230372)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.001339286314449, pvalue=0.3166628057230372)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=7.250e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=3.819e-01 stat=5.880e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=6.610e+02\n",
      "0.6918880692040176 0.5578990670070274\n",
      "CEBPE_2\n",
      "0h vs 4h RanksumsResult(statistic=0.9689875342049256, pvalue=0.3325514088841939)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.64639708117969, pvalue=0.09968203363622569)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.64639708117969, pvalue=0.09968203363622569)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=6.538e-01 stat=1.463e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=7.916e-03 stat=1.256e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=1.572e+04\n",
      "2.529955662869097 0.05590708984675098\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_26765/3605437427.py:81: RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`). Consider using `matplotlib.pyplot.close()`.\n",
      "  plt.figure(figsize=(4.5,6))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALOX5AP\n",
      "0h vs 4h RanksumsResult(statistic=-1.06993006993007, pvalue=0.28465078680354083)\n",
      "0h vs 2d  RanksumsResult(statistic=-0.567649741562785, pvalue=0.5702728243961538)\n",
      "0h vs 4d  RanksumsResult(statistic=-0.567649741562785, pvalue=0.5702728243961538)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=3.977e-01 stat=1.431e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=1.536e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=9.403e-01 stat=1.489e+04\n",
      "0.4669970548445846 0.7053677133801486\n",
      "CEBPB_1\n",
      "0h vs 4h RanksumsResult(statistic=0.6085863259104622, pvalue=0.5427986703665875)\n",
      "0h vs 2d  RanksumsResult(statistic=-2.8199061372444687, pvalue=0.004803769860372517)\n",
      "0h vs 4d  RanksumsResult(statistic=-2.8199061372444687, pvalue=0.004803769860372517)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=7.555e-01 stat=3.879e+05\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=2.361e-04 stat=3.525e+05\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=2.777e-01 stat=3.816e+05\n",
      "3.3968004539051577 0.017086838766668735\n",
      "CEBPB_2\n",
      "0h vs 4h RanksumsResult(statistic=-1.0356688544225388, pvalue=0.3003566573741131)\n",
      "0h vs 2d  RanksumsResult(statistic=-2.5354585967992405, pvalue=0.011230020269255486)\n",
      "0h vs 4d  RanksumsResult(statistic=-2.5354585967992405, pvalue=0.011230020269255486)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=4.184e-01 stat=3.762e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=9.999e-04 stat=3.265e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=6.999e-01 stat=3.830e+04\n",
      "1.2297165843173619 0.2974163996389942\n",
      "CEBPB_3\n",
      "0h vs 4h RanksumsResult(statistic=-0.8190774512218213, pvalue=0.41274222596224264)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.778847817888765, pvalue=0.07526471514528636)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.778847817888765, pvalue=0.07526471514528636)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=8.943e-01 stat=1.462e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=9.726e-02 stat=1.335e+04\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=6.133e-02 stat=1.314e+04\n",
      "1.1608193829738331 0.32360542839371553\n",
      "IRF5\n",
      "0h vs 4h RanksumsResult(statistic=-0.21222414724873398, pvalue=0.8319321690027655)\n",
      "0h vs 2d  RanksumsResult(statistic=-0.28396188716379894, pvalue=0.7764395884834305)\n",
      "0h vs 4d  RanksumsResult(statistic=-0.28396188716379894, pvalue=0.7764395884834305)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=7.080e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=7.500e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=7.090e+02\n",
      "0.14906435889381187 0.9302083403982924\n",
      "CEBPA\n",
      "0h vs 4h RanksumsResult(statistic=-0.6899593872424348, pvalue=0.49021972757978005)\n",
      "0h vs 2d  RanksumsResult(statistic=-1.287117219475653, pvalue=0.19805343446537438)\n",
      "0h vs 4d  RanksumsResult(statistic=-1.287117219475653, pvalue=0.19805343446537438)\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4h: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=4.410e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 2d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=9.039e-02 stat=3.260e+02\n",
      "Using zero_method  wilcox\n",
      "0h v.s. 4d: Wilcoxon test (paired samples) with Bonferroni correction, P_val=1.000e+00 stat=4.460e+02\n",
      "0.8580311063387155 0.4640689317007203\n"
     ]
    }
   ],
   "source": [
    "\n",
    "## save all the figures in a loop to one pdf file. \n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "pdf = PdfPages(\"./Gene_asso_region_ChrInt.pdf\")\n",
    "\n",
    "\n",
    "for Gene in  hubs_anchor[\"Gene\"]:\n",
    "    print(Gene)\n",
    "    \n",
    "    df = pd.DataFrame({\"0h\" : Hic_contact_list[0][Gene],\n",
    "                     \"4h\" : Hic_contact_list[1][Gene],\n",
    "                     \"2d\" : Hic_contact_list[2][Gene],\n",
    "                      \"4d\" : Hic_contact_list[3][Gene]})\n",
    "    \n",
    "    df.to_csv(\"./\" + Gene + \"region_ChrInt.csv\") \n",
    "    \n",
    "    df2 = pd.melt(df,var_name='Condition', value_name='ChrInt')\n",
    "    \n",
    "    # draw boxplot\n",
    "\n",
    "    ## chnage boxplot color:\n",
    "    PROPS = {\n",
    "        'boxprops':{'facecolor':'none', 'edgecolor':'black'},\n",
    "        'medianprops':{'color':'black'},\n",
    "        'whiskerprops':{'color':'black', 'linestyle':'--'},\n",
    "        'capprops':{'color':'black'}\n",
    "    }\n",
    "    # 1. horizontal ---- but hard to add p value \n",
    "    plt.figure(figsize=(6,4.5))\n",
    "    ax1 = sns.boxplot(x='ChrInt',y='Condition', data=df2, color = \"white\",width = 0.4,showfliers=False,**PROPS)\n",
    "    sns.stripplot(x='ChrInt',y='Condition', data=df2, color = \"none\", dodge=True,size = 2,jitter = 0.15,ax = ax1 )\n",
    "\n",
    "    ax1.set_title(Gene)\n",
    "\n",
    "\n",
    "    # ---------Remove the top and right spines from plot(s)---------\n",
    "    sns.despine(fig=None, ax=None, top=True, right=True, left=False, bottom=False, offset=None, trim=False)\n",
    "\n",
    "    # plt.show()\n",
    "    \n",
    "\n",
    "    pvalue1 = scipy.stats.ranksums(df2[df2.Condition == \"0h\"][\"ChrInt\"].dropna(),df2[df2.Condition == \"4h\"][\"ChrInt\"].dropna())\n",
    "    print( \"0h vs 4h \"  + str( pvalue1))\n",
    "\n",
    "    pvalue2 = scipy.stats.ranksums(df2[df2.Condition == \"0h\"][\"ChrInt\"].dropna(),df2[df2.Condition == \"2d\"][\"ChrInt\"].dropna())\n",
    "    print( \"0h vs 2d  \"  + str( pvalue2))\n",
    "    \n",
    "    pvalue3 = scipy.stats.ranksums(df2[df2.Condition == \"0h\"][\"ChrInt\"].dropna(),df2[df2.Condition == \"4d\"][\"ChrInt\"].dropna())\n",
    "    print( \"0h vs 4d  \"  + str( pvalue2))\n",
    "    \n",
    "    pdf.savefig() \n",
    "\n",
    "    \n",
    "    # 2. Vertical ----easy to add p value\n",
    "    plt.figure(figsize=(4.5,6))\n",
    "    \n",
    "    ax2 = sns.boxplot(x='Condition', y='ChrInt', data=df2, color = \"white\",width = 0.4,showfliers=False,**PROPS)\n",
    "    sns.stripplot(x='Condition', y='ChrInt', data=df2, color = \"none\", dodge=True,size = 2,jitter = 0.15,ax = ax2)\n",
    "    \n",
    "    \n",
    "    ax2.set_title(Gene)\n",
    "    # MOVE Y AXIS TO THE RIGHT \n",
    "    ax2.yaxis.set_label_position(\"right\")\n",
    "    ax2.yaxis.tick_right()\n",
    "    # ---------Remove the top andLEFT spines from plot(s)---------\n",
    "    sns.despine(fig=None, ax=None, top=True, right=False, left= True, bottom=False, offset=None, trim=False)\n",
    "\n",
    "    from statannot import add_stat_annotation\n",
    "\n",
    "\n",
    "    add_stat_annotation(ax2, data=df2,x='ChrInt',y='Condition',\n",
    "                        box_pairs=[(\"0h\", \"4h\"), (\"0h\", \"2d\"),(\"0h\", \"4d\")],\n",
    "                        test='Wilcoxon', text_format='full', loc='inside')\n",
    "    \n",
    "    pdf.savefig() \n",
    "    \n",
    "    # 3. anova test\n",
    "    plt.figure(figsize=(4.5,6))\n",
    "    \n",
    "    ax2 = sns.boxplot(x='Condition', y='ChrInt', data=df2, color = \"white\",width = 0.4,showfliers=False,**PROPS)\n",
    "    sns.stripplot(x='Condition', y='ChrInt', data=df2, color = \"none\", dodge=True,size = 2,jitter = 0.15,ax = ax2)\n",
    "    \n",
    "    # ax2.set_title(Gene)\n",
    "    # MOVE Y AXIS TO THE RIGHT \n",
    "    ax2.yaxis.set_label_position(\"right\")\n",
    "    ax2.yaxis.tick_right()\n",
    "    # ---------Remove the top andLEFT spines from plot(s)---------\n",
    "    sns.despine(fig=None, ax=None, top=True, right=False, left= True, bottom=False, offset=None, trim=False)\n",
    "\n",
    "    # stats f_oneway functions takes the groups as input and returns ANOVA F and p value\n",
    "    import scipy.stats as stats\n",
    "    fvalue, pvalue = stats.f_oneway(df['0h'], df['4h'], df['2d'], df['4d'])\n",
    "    print(fvalue, pvalue)\n",
    "    significance = \"*\" if pvalue < 0.05 else \"ns\"\n",
    "    \n",
    "    T= Gene+\" \\n\"+ f'ANOVA p-value: {pvalue:.4f} {significance}'\n",
    "    ax2.set_title(T)\n",
    "\n",
    "    pdf.savefig() \n",
    "\n",
    "    \n",
    "pdf.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
