---
title: "2_Fig2_FigS3"
output: 
  html_document:
    number_sections: true
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
      smooth_scroll: false
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "./RNAseq/HL60/")

```

```{r}
rm(list = ls())
PATH <- "./RNAseq/HL60/"
```

## load library
```{r}
library(ggplot2)
library(ggrepel)
library(stringr)
library(DESeq2)
library(edgeR)
library(grid)
library(factoextra)
library(cluster)
library(pbapply)
library(pheatmap)
library(openxlsx)
library(ggrepel)
library(GenomicFeatures)
library(GenomicRanges)
# library("org.Rn.eg.db") # rat
library("org.Hs.eg.db") # human
library("org.Mm.eg.db") # mm

library("biomaRt")
library(clusterProfiler)
library(openxlsx)
library(rtracklayer)
library(plyr) # to use ioin
library(dplyr)# use full_join
library(gtools) # mixsort
library(grid)
library(futile.logger)
library(VennDiagram)
```

```{r}
plot_theme <- theme_bw() +
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        # title in the middle
        plot.title = element_text(hjust = 0.5),
        text = element_text(size = 15),axis.text = element_text(size = 15)) +theme(aspect.ratio = 1)
```


## preparation: get the exon length from gtf files 

### hg19

prepare the file beforehand 
```{r}

exonLen <- read.table("./RNAseq/exonLen.txt")
dim(exonLen)
head(exonLen)

```

## calculate TPM and PCA plot

### set sample order
```{r}
## set sample order
count_files <-dir("htseq_count_files") 
count_files <- count_files[grep(".count",count_files)]
length(count_files)

extract_sample <- function(x) {return  (unlist(strsplit(x, "[.]"))[1])}
samples <- unlist(lapply(count_files,extract_sample))
sample_order<- mixedsort(samples)
# 4h in the front 
sample_order <- sample_order[c(1:3,10:12,4:9)]

length(sample_order)
sample_order
```

### import htseq counts
```{r}
# input rawcount
CountList=list()


for (S in sample_order){

  temp <- read.table(paste0("htseq_count_files/",count_files[grep(S,count_files)]),header = F)
  tail(temp) 
  temp <- temp[1:(dim(temp)[1]-5),]
  
  colnames(temp) <- c("Gene",S)
 
  head(temp)
  CountList[[S]]<- temp
}

RawCounts <- Reduce(full_join,CountList)
dim(RawCounts )  

write.csv(RawCounts,file = "RawCounts.csv",row.names = FALSE )
tail(RawCounts)
```


```{r}
exonLen <- read.table("./RNAseq/exonLen.txt")
colnames(exonLen) <- c("Gene","Exon_length")
head(exonLen)

coverage <- read.csv(paste0(PATH,"RawCounts.csv"),header = T)
# name <- read.csv(paste0(PATH,"name.txt"),header = F)
head(coverage)
coverage <- coverage[order(coverage$Gene),] # then the rownames of coverage and rownames of 

coverage  <- merge(x = exonLen,y = coverage,by = "Gene")
head(coverage)
dim(exonLen)[1] == dim(coverage)[1]
rownames(coverage) <- coverage$Gene

names(coverage)
```

### calculate TPM 
```{r}
RPK <- coverage[,3:dim(coverage)[2]]/coverage$Exon_length*10^3

TPM <- as.data.frame(t(t(RPK)/apply(RPK, 2, sum)*10^6))

dim(TPM)
colnames(TPM) <- paste0(colnames(TPM) ,"_TPM")
TPM[["Gene"]] <- coverage$Gene
TPM <- TPM[,c(dim(TPM)[2],1:(dim(TPM)[2]-1))]

head(TPM)

length(which((rownames(coverage)==rownames(TPM))==TRUE))

write.csv(TPM,paste0(PATH,'TPM.csv'))


```


### PCA all samples
```{r}

TPM_pca <- TPM[2:(dim(TPM)[2])][which(apply(TPM[,2:(dim(TPM)[2])], 1, var)!=0),]
dim(TPM_pca)

pca <- as.data.frame(prcomp(t(TPM_pca),scale.=T)$x)
pca$name <- rownames(pca)

fun1 <- function(x) {return (substr(x,1,gregexpr(pattern ='_',x)[[1]][1]-1))}
 
pca$sample <- (unlist(lapply(sample_order,fun1)))

PCA <- prcomp(t(TPM_pca),scale.=T);PC1=(PCA$sdev/sum(PCA$sdev))[1];PC2=(PCA$sdev/sum(PCA$sdev))[2]

ggplot(data = pca,aes(x=PC1, y=PC2, label=name, colour = sample)) + theme_bw() +
  geom_point(size = 3) +  geom_text_repel(aes(label=name), size = 4, hjust=0.2, vjust=0.5, angle=0) +
  labs(x = paste('PC1 (',round(PC1,3)*100,'% variance)',sep=''),y=paste('PC2 (',round(PC2,3)*100,'% variance)',sep='')) +
  theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank())

```
### remove replicate based on PCA plot 


RA4h_1 shoule be removed according to the PCA plot
```{r}
sample_order <- sample_order[-which(sample_order == "RA4h_1")]
sample_order


coverage<-coverage[,-which(colnames(coverage)=="RA4h_1")] 

TPM<-TPM[,-which(colnames(TPM)=="RA4h_1_TPM")] 
```
### PCA all samples after remove outlier replicate

```{r}
names(TPM_pca)

TPM_pca <- TPM[2:(dim(TPM)[2])][which(apply(TPM[,2:(dim(TPM)[2])], 1, var)!=0),]
dim(TPM_pca)

dim(TPM_pca)

pca <- as.data.frame(prcomp(t(TPM_pca),scale.=T)$x)
pca$name <- rownames(pca)

fun1 <- function(x) {return (substr(x,1,gregexpr(pattern ='_',x)[[1]][1]-1))}
 
pca$sample <- (unlist(lapply(sample_order,fun1)))

PCA <- prcomp(t(TPM_pca),scale.=T);PC1=(PCA$sdev/sum(PCA$sdev))[1];PC2=(PCA$sdev/sum(PCA$sdev))[2]

ggplot(data = pca,aes(x=PC1, y=PC2, label=name, colour = sample)) + theme_bw() +
  geom_point(size = 3) +  geom_text_repel(aes(label=name), size = 4, hjust=0.2, vjust=0.5, angle=0) +
  labs(x = paste('PC1 (',round(PC1,3)*100,'% variance)',sep=''),y=paste('PC2 (',round(PC2,3)*100,'% variance)',sep='')) +
  theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank())


```


## find DEG ---heatmap on all clusters--output files

```{r}
dir.create(paste0(PATH,"Filter_DEG"))

# diff <- numeric();
N = length(unique(pca$sample))

Conditions <- unique(unlist(lapply(sample_order,fun1)))

for (i in 1:(N-1)){
  for (j in (i+1):N){
    # i = 1
    # j = 4
    condition1 = grep(colnames(coverage),pattern = Conditions[i])
    condition2 = grep(colnames(coverage),pattern = Conditions[j])
    
    group <- factor(c(rep(i,length(condition1)),rep(j,length(condition2))))
  
    y = DGEList(counts = coverage[,c(condition1,condition2)],group = group)
    print(names(coverage[,c(condition1,condition2)]))
    dim(y)
    
    ## filter y to remove low expression genes
    keep.exprs <- filterByExpr(y, group=group)
    y <- y[keep.exprs,]
    dim(y)
     ##  ##  ##  ##  ##  ##  ##  ##  ##  ## 
    
    y <- calcNormFactors(y)
    y <- estimateCommonDisp(y)
    fit <- glmQLFit(y)
    qlf <- glmQLFTest(fit)
    fdr <- p.adjust(p = qlf$table$PValue,method = "BH",n = length(qlf$table$PValue))
    length(fdr)
    diff_idx <- which(fdr < 0.05 & abs(qlf$table$logFC) > log2(2))

    pair_name <- paste(Conditions[j],'_vs_',Conditions[i],sep='')

    ##  after filter, the dim will be different, then you need to add gene to the change 
    change <- data.frame(Gene = rownames(qlf$table),logFC=qlf$table$logFC,pval=qlf$table$PValue,fdr=fdr)
     ##---##--##--##--##--##---##--##--##--##--
    
    colnames(change) <- c("Gene",paste(pair_name,'(logFC)'),paste(pair_name,'(pval)'),paste(pair_name,'(fdr)'))
   
    ########################### save the data ###############################################################################
    dim(change)[1] == dim(coverage)[1] 
    dim(coverage)[1] ==dim(TPM)[1]

    #  add raw count
    change  <- merge(x = change,y = coverage[,c(1,condition1,condition2)],by = "Gene",all.x = T)
    dim(change)
    
    # add TPM 
    change  <- merge(x = change,y = TPM[,c(1,grep(colnames(TPM),pattern = Conditions[i]),grep(colnames(TPM),pattern = Conditions[j]))],by = "Gene",all.x = T)
    dim(change)
    
    head(change)
  
    # row.names = T to keep gene name 
    write.csv(change,paste0(PATH,"Filter_DEG/",Conditions[i],'_',Conditions[j],'.csv',sep=''),
               quote = F,row.names = T)
    
    ### (2) save the DEGs -------------------

    # significant  change here -----
    change_sig<- change[diff_idx,]
    colnames(change_sig)
    dim(change_sig)
    write.csv(change_sig,paste(PATH,"Filter_DEG/",Conditions[i],'_',Conditions[j],'_DEG.csv',sep=''),quote = F,row.names = T,col.names = T)
    # previously, diff_index already been filtered: diff_idx <- which(fdr < 0.05 & abs(qlf$table$logFC) > log2(2))
    
    write.csv(change_sig[change_sig[,grep(colnames(change_sig),pattern = "logFC" )]>0,],paste(PATH,"Filter_DEG/",Conditions[i],'-less-than-',Conditions[j],'.csv',sep=''),
                quote = F,row.names = T,col.names = T)
    print(paste0(Conditions[i],'-less-than-',Conditions[j],": ",dim(change_sig[change_sig[,grep(colnames(change_sig),pattern = "logFC" )]>0,])[1]))
    
    write.csv(change_sig[change_sig[,grep(colnames(change_sig),pattern = "logFC" )]<0,],paste(PATH,"Filter_DEG/",Conditions[i],'-more-than-',Conditions[j],'.csv',sep=''),
                quote = F,row.names = T,col.names = T)
    print(paste0(Conditions[i],'-more-than-',Conditions[j],": ",dim(change_sig[change_sig[,grep(colnames(change_sig),pattern = "logFC" )]<0,])[1]))
    
    
    ########################### ########################### ########################### ###########################

    # heat map on all the samples -------------  hcluster
    # Pay attention , diff_index ia based on the filterd data !!!!!!!!! 
    DEG <- change$Gene[unique(diff_idx)]
    
    length(DEG)

    samples <- as.data.frame(pca$sample); names(samples) <- 'sample';
    
    # rownames(samples) <- pca$name;
    rownames(samples) <- colnames(TPM)[2:length(colnames(TPM))] # make sure the same is consistent, because I changed the colnames of TPM previously
    samples$sample <- factor(samples$sample)
    samples

    mtx_Zscore <- (TPM[DEG,2:(dim(TPM)[2])]-apply(X = TPM[DEG,2:(dim(TPM)[2])], MARGIN = 1, mean))/apply(X = TPM[DEG,2:(dim(TPM)[2])], MARGIN = 1, sd)
    # remove NA or inf value  
      # rows that should kept 
      row_keep <- apply(mtx_Zscore, 1, function(x) !(sum(is.na(x)) | sum(is.infinite(x))) )
      mtx_Zscore = mtx_Zscore[row_keep, ]

    mtx_hclust <- hclust(dist(mtx_Zscore), method = "complete")

    gene_col <- data.frame(cutree(tree = (mtx_hclust), k = 2))
    names(gene_col)<- 'cluster';gene_col$cluster <- factor(gene_col$cluster)

    pheatmap(mtx_Zscore,cluster_rows=T,cluster_cols = F,show_rownames = F,cutree_rows=2,
             angle_col='315',annotation_row = gene_col, annotation_col = samples,main = paste0(' DEGs z-score (hclust)',Conditions[i],'-VS-',Conditions[j]))
    

    fviz_nbclust(mtx_Zscore, kmeans, method = "wss")

    set.seed(2);mtx_kmeans <- kmeans(mtx_Zscore,centers=2,nstart = 25)
    gene_kmeans <- as.data.frame(sort(mtx_kmeans$cluster)); names(gene_kmeans) <- 'cluster'; 
    gene_kmeans$cluster <- factor(gene_kmeans$cluster)
    
    pheatmap(mtx_Zscore[names(sort(mtx_kmeans$cluster)),],cluster_rows=F,cluster_cols = F,show_rownames = F,
             angle_col='315',annotation_row = gene_kmeans, annotation_col = samples,main = 'DEGs z-score (k-means)')
  
  }
}

```

## find DEG --- heatmap on paired clusters---volcano plot


```{r}
# diff <- numeric(); 
N = length(unique(pca$sample))
Conditions <- unique(unlist(lapply(sample_order,fun1)))

for (i in 1:(N-1)){
  for (j in (i+1):N){

    condition1 = grep(colnames(coverage),pattern = Conditions[i])
    condition2 = grep(colnames(coverage),pattern = Conditions[j])
    group <- factor(c(rep(i,length(condition1)),rep(j,length(condition2))))
    y = DGEList(counts = coverage[,c(condition1,condition2)],group = group)
    print(names(coverage[,c(condition1,condition2)]))
    

    keep.exprs <- filterByExpr(y, group=group)
    y <- y[keep.exprs,]
    dim(y)

      y <- calcNormFactors(y)
      y <- estimateCommonDisp(y)
      fit <- glmQLFit(y)
      qlf <- glmQLFTest(fit)
      fdr <- p.adjust(p = qlf$table$PValue,method = "BH",n = length(qlf$table$PValue))
      diff_idx <- which(fdr < 0.05 & abs(qlf$table$logFC) > log2(2))

       pair_name <- paste(Conditions[j],'_vs_',Conditions[i],sep='')

        change <- data.frame(Gene = rownames(qlf$table),logFC=qlf$table$logFC,pval=qlf$table$PValue,fdr=fdr)

        colnames(change) <- c("Gene",paste(pair_name,'(logFC)'),paste(pair_name,'(pval)'),paste(pair_name,'(fdr)'))
        dim(change)
        
      #########################################################################################

      # heat map on paired samples -----hcluster 
      
      DEG <- change$Gene[unique(diff_idx)]
      paired_sample <- names(coverage[,c(condition1,condition2)])
      # fun2 <- function(x) {return (substr(x,1,str_length(x)-1))}
      fun1 <- function(x) {return (substr(x,1,gregexpr(pattern ='_',x)[[1]][1]-1))}
      paired_group <-  unlist(lapply(paired_sample,fun1))
      samples <- as.data.frame(paired_group); names(samples) <- 'sample';
    
      rownames(samples) <- paste0(paired_sample,"_TPM"); # change here to keep the name consistent
      samples$sample <- factor(samples$sample)
     
      samples
      
      mtx_Zscore <- (TPM[DEG,paste0(paired_sample,"_TPM")]-apply(X = TPM[DEG,paste0(paired_sample,"_TPM")], MARGIN = 1, mean))/apply(X = TPM[DEG,paste0(paired_sample,"_TPM")], MARGIN = 1, sd)

      row_keep <- apply(mtx_Zscore, 1, function(x) !(sum(is.na(x)) | sum(is.infinite(x))) )
      mtx_Zscore = mtx_Zscore[row_keep, ]
      mtx_hclust <- hclust(dist(mtx_Zscore), method = "complete")

      gene_col <- data.frame(cutree(tree = (mtx_hclust), k = 2))
      names(gene_col)<- 'cluster';gene_col$cluster <- factor(gene_col$cluster)
      pheatmap(mtx_Zscore,cluster_rows=T,cluster_cols = F,show_rownames = F,cutree_rows=2,
               angle_col='315',annotation_row = gene_col, annotation_col = samples,main = paste0('Pair-wise DEGs z-score (hclust)',Conditions[i],'-VS-',Conditions[j]))
      
      
      ##  heat map on paired samples ----------kmeans
      fviz_nbclust(mtx_Zscore, kmeans, method = "wss")
  
      set.seed(2);mtx_kmeans <- kmeans(mtx_Zscore,centers=2,nstart = 25)
      gene_kmeans <- as.data.frame(sort(mtx_kmeans$cluster)); names(gene_kmeans) <- 'cluster'; 
      gene_kmeans$cluster <- factor(gene_kmeans$cluster)
      
      pheatmap(mtx_Zscore[names(sort(mtx_kmeans$cluster)),],cluster_rows=F,cluster_cols = F,show_rownames = F,angle_col='315',annotation_row = gene_kmeans, annotation_col = samples,main = 'Pair-wise DEGs z-score (k-means)')
      
      ########################################################################################
      ## plot the DEGs using volcano plot 

      change <- data.frame(Gene = rownames(qlf$table),logFC=qlf$table$logFC,pval=qlf$table$PValue,fdr=fdr)
      change[["sig"]] <- NA
      change[["sig"]][diff_idx] <- "Yes"
      change[["sig"]][-diff_idx] <- "No"
      change[["gene"]] <- change$Gene
      head(change) 
      
      p<- ggplot(data=change, aes(x= logFC , y=-log10(fdr))) + geom_point(col = ifelse(change$sig=="No","grey",ifelse(change$logFC>0,"red","blue")), size = ifelse(change$sig=="Yes",1,0.1)) + 
        labs(y = "-log10(fdr)", x = paste(pair_name,"logFC"),title = paste0(Conditions[i],'-VS-',Conditions[j],":  DEGs(qval < 0.05; FC > 2)"))+
         ylim(c(0, 10))+
        # add line according to the threshold use previously
          geom_hline(yintercept = -log10(0.05),lty=4,col="black",lwd=0.6)+
          geom_vline(xintercept = log2(2),lty=4,col="black",lwd=0.6)+
          geom_vline(xintercept = -log2(2),lty=4,col="black",lwd=0.6)+plot_theme
      
      print(p)
 
      # no ylim
      p1 <- ggplot(data=change, aes(x= logFC , y=-log10(fdr))) + geom_point(col = ifelse(change$sig=="No","grey",ifelse(change$logFC>0,"red","blue")), size = ifelse(change$sig=="Yes",1,0.1)) + 
        labs(y = "-log10(fdr)", x = paste(pair_name,"logFC"),title = paste0(Conditions[i],'-VS-',Conditions[j],":  DEGs(qval < 0.05; FC > 2)"))+
        # add line according to the threshold use previously
          geom_hline(yintercept = -log10(0.05),lty=4,col="black",lwd=0.6)+
          geom_vline(xintercept = log2(2),lty=4,col="black",lwd=0.6)+
          geom_vline(xintercept = -log2(2),lty=4,col="black",lwd=0.6)+plot_theme
      
      ## label the top changed genes!!!------------------
      p1_label_gene<- ggplot(data=change, aes(x= logFC , y=-log10(fdr))) + geom_point(col = ifelse(change$sig=="No","grey",ifelse(change$logFC>0,"red","blue")), size = ifelse(change$sig=="Yes",1,0.1)) + 
        labs(y = "-log10(fdr)", x = paste(pair_name,"logFC"),title = paste0(Conditions[i],'-VS-',Conditions[j],":  DEGs(qval < 0.05; FC > 2)"))+
        # add line according to the threshold use previously
          geom_hline(yintercept = -log10(0.05),lty=4,col="black",lwd=0.6)+
          geom_vline(xintercept = log2(2),lty=4,col="black",lwd=0.6)+
          geom_vline(xintercept = -log2(2),lty=4,col="black",lwd=0.6)+
        geom_text_repel(aes(label=ifelse((logFC>5|logFC< -5),gene,"")))+plot_theme
      
  print( p1_label_gene)
      ## compress y axis again
      p2_label_gene <-  ggplot(data=change, aes(x= logFC , y=log10(-log10(fdr)+1))) + geom_point(col = ifelse(change$sig=="No","grey",ifelse(change$logFC>0,"red","blue")), size = ifelse(change$sig=="Yes",1,0.1)) + 
        labs(y = "log10(-log10(fdr)+1)", x = paste(pair_name,"logFC"),title = paste0(Conditions[i],'-VS-',Conditions[j],":  Diff Genes(qval < 0.05; FC > 2)"))+
        # add line according to the threshold use previously
        geom_hline(yintercept = log10(-log10(0.05)+1),lty=4,col="black",lwd=0.6)+
        geom_vline(xintercept = log2(2),lty=4,col="black",lwd=0.6)+
        geom_vline(xintercept = -log2(2),lty=4,col="black",lwd=0.6)+
        geom_text_repel(aes(label=ifelse((logFC>5|logFC< -5),gene,"")))+plot_theme
      
       print(p2_label_gene)
       
       ## add FC box plot
       p <- ggplot(data=change, aes(y= logFC)) + geom_boxplot(notch = T)+plot_theme+labs(title = paste0(Conditions[i],'-VS-',Conditions[j]))+ ylim(-5,5)+theme(aspect.ratio = 1)+plot_theme
       print(p)
       
       plot(density(change$logFC,na.rm=T),main = paste0(Conditions[i],'-VS-',Conditions[j]))
       

  }
}

```


## ---- other specific exploration

### overlap among all the DEGs 

```{r}
# input rawcount
DEG_List=list()

DEG_file<- dir("Filter_DEG/")[grep(dir("Filter_DEG/"),pattern = "_DEG")]

#  1. ------------filter IRI : 0 < gene_IRI & gene_IRI<1  

for (f in DEG_file){
  # print(S)
  # f <- DEG_file[1]
  # temp_all <- read.table(paste0("IR_",files[grep(glob2rx(paste0("*",S,".quant.IRI.genes.txt")),files)]),header = TRUE)
  temp <- read.csv(paste0("Filter_DEG/",f))[,-1]
  tail(temp) 
  name <- strsplit(f,split = "_DEG.csv")[[1]]
  DEG_List[[name]]<- temp$Gene
}

names(DEG_List)

lapply(DEG_List, length)
```

### upset 
```{r}
library(UpSetR)
listInput <- list(Ctrl_RA4h=DEG_List[["Ctrl_RA4h"]],Ctrl_RA2d=DEG_List[["Ctrl_RA2d"]],Ctrl_RA4d=DEG_List[["Ctrl_RA4d"]],RA4h_RA2d=DEG_List[["RA4h_RA2d"]],RA4h_RA4d=DEG_List[["RA4h_RA4d"]],RA2d_RA4d=DEG_List[["RA2d_RA4d"]])


upset(fromList(listInput), order.by = "freq",nsets = 6)

listInput <- list(Ctrl_RA4h=DEG_List[["Ctrl_RA4h"]],Ctrl_RA2d=DEG_List[["Ctrl_RA2d"]],Ctrl_RA4d=DEG_List[["Ctrl_RA4d"]])
upset(fromList(listInput), order.by = "freq")

listInput <- list(RA4h_RA2d=DEG_List[["RA4h_RA2d"]],RA4h_RA4d=DEG_List[["RA4h_RA4d"]])
upset(fromList(listInput), order.by = "freq")

```

### venn plot 

```{r}

listInput <- list(Ctrl_RA4h=DEG_List[["Ctrl_RA4h"]],Ctrl_RA2d=DEG_List[["Ctrl_RA2d"]],Ctrl_RA4d=DEG_List[["Ctrl_RA4d"]])

v0 <- venn.diagram(listInput , filename="Filter_DEG/DEG_venn_3_ctrl", 
                    fill = c("red", "blue", "green"),
                    alpha = 0.50,
                   # number size and lable size
                   cat.cex = 2,cex = 2,
                   fontface = "bold",cat.fontface = "bold",cat.default.pos = "outer",
                    margin = 0.2, # Number giving the amount of whitespace around the diagram
                   col = "transparent")
overlaps <- calculate.overlap(listInput)
# grid.draw(v0)

```


```{r}

## at most 5 list
listInput <- list(RA4h_RA2d=DEG_List[["RA4h_RA2d"]],RA4h_RA4d=DEG_List[["RA4h_RA4d"]])


v0 <- venn.diagram(listInput , filename="Filter_DEG/DEG_venn_2_4h", 
                    fill = c("red", "blue"),
                    alpha = 0.50,
                   # number size and lable size
                   cat.cex = 2,cex = 2,
                   fontface = "bold",cat.fontface = "bold",cat.default.pos = "outer",
                   margin = 0.2, # Number giving the amount of whitespace around the diagram
                    col = "transparent")
overlaps <- calculate.overlap(listInput)
# grid.draw(v0)

```


```{r}

## at most 5 list
listInput <- list(Ctrl_RA2d=DEG_List[["Ctrl_RA2d"]],Ctrl_RA4d=DEG_List[["Ctrl_RA4d"]],RA4h_RA2d=DEG_List[["RA4h_RA2d"]],RA4h_RA4d=DEG_List[["RA4h_RA4d"]],RA2d_RA4d=DEG_List[["RA2d_RA4d"]])

v0 <- venn.diagram(listInput , filename="Filter_DEG/DEG_venn_5", 
                    fill = c("red", "blue", "green","grey","yellow"),
                    alpha = 0.50,
                   # number size and lable size
                   cat.cex = 1,cex = 1,
                   fontface = "bold",cat.fontface = "bold",cat.default.pos = "outer",
                    margin = 0.2, # Number giving the amount of whitespace around the diagram
                   col = "transparent")
overlaps <- calculate.overlap(listInput)
# grid.draw(v0)

```


### save all_DEG_TPM
```{r}
All_DEGs <- unique(unlist(DEG_List))

head(TPM)

All_DEGs_TPM <- TPM[which(TPM$Gene %in% All_DEGs),]
dim(All_DEGs_TPM)

write.csv(All_DEGs_TPM,"Filter_DEG/All_DEGs_TPM.csv")
```


### overall heatmap of all the DEGs (TPM)


```{r}
    All_DEGs <- unique(unlist(DEG_List))
  # heat map on all the samples -------------hcluster
    samples <- as.data.frame(pca$sample); names(samples) <- 'sample';
    
    # rownames(samples) <- pca$name;
    rownames(samples) <- colnames(TPM)[2:length(colnames(TPM))] 
    samples$sample <- factor(samples$sample)
    samples

    mtx_Zscore <- (TPM[All_DEGs,2:(dim(TPM)[2])]-apply(X = TPM[All_DEGs,2:(dim(TPM)[2])], MARGIN = 1, mean))/apply(X = TPM[All_DEGs,2:(dim(TPM)[2])], MARGIN = 1, sd)
    # remove NA or inf value  
      # rows that should kept 
    row_keep <- apply(mtx_Zscore, 1, function(x) !(sum(is.na(x)) | sum(is.infinite(x))) )
    mtx_Zscore = mtx_Zscore[row_keep, ]
    
    mtx_hclust <- hclust(dist(mtx_Zscore), method = "complete")
  
    gene_col <- data.frame(cutree(tree = (mtx_hclust), k = 2))
    names(gene_col)<- 'cluster';gene_col$cluster <- factor(gene_col$cluster)

    pheatmap(mtx_Zscore,cluster_rows=T,cluster_cols = F,show_rownames = F,cutree_rows=2,
             angle_col='315',annotation_row = gene_col, annotation_col = samples,main = 'All DEGs z-score (hclust)')
    
    ##  heat map on all the samples --------------kmeans
    dim(mtx_Zscore)
    
    fviz_nbclust(mtx_Zscore, kmeans, method = "wss")

    set.seed(2);mtx_kmeans <- kmeans(mtx_Zscore,centers=4,nstart = 25)
    gene_kmeans <- as.data.frame(sort(mtx_kmeans$cluster)); names(gene_kmeans) <- 'cluster'; 
    gene_kmeans$cluster <- factor(gene_kmeans$cluster)
    
    pheatmap(mtx_Zscore[names(sort(mtx_kmeans$cluster)),],cluster_rows=F,cluster_cols = F,show_rownames = F,
             angle_col='315',annotation_row = gene_kmeans, annotation_col = samples,main = 'All DEGs z-score (k-means)')
    # names(mtx_kmeans$cluster)
    Cluster_4_list <- list()
    Cluster_4_list[["Cluster1"]]<- names(mtx_kmeans$cluster[which(mtx_kmeans$cluster==1)])
    Cluster_4_list[["Cluster2"]]<- names(mtx_kmeans$cluster[which(mtx_kmeans$cluster==2)])
    Cluster_4_list[["Cluster3"]]<- names(mtx_kmeans$cluster[which(mtx_kmeans$cluster==3)])
    Cluster_4_list[["Cluster4"]]<- names(mtx_kmeans$cluster[which(mtx_kmeans$cluster==4)])
    lapply(Cluster_4_list, length)
    
    
    set.seed(2);mtx_kmeans <- kmeans(mtx_Zscore,centers=2,nstart = 25)
    gene_kmeans <- as.data.frame(sort(mtx_kmeans$cluster)); names(gene_kmeans) <- 'cluster'; 
    gene_kmeans$cluster <- factor(gene_kmeans$cluster)
    
    pheatmap(mtx_Zscore[names(sort(mtx_kmeans$cluster)),],cluster_rows=F,cluster_cols = F,show_rownames = F,
             angle_col='315',annotation_row = gene_kmeans, annotation_col = samples,main = 'All DEGs z-score (k-means)')
  
    Cluster_2_list <- list()
    Cluster_2_list[["Cluster1"]]<- names(mtx_kmeans$cluster[which(mtx_kmeans$cluster==1)])
    Cluster_2_list[["Cluster2"]]<- names(mtx_kmeans$cluster[which(mtx_kmeans$cluster==2)])

    lapply(Cluster_2_list, length)
    
```

#### save the genes of the two clusters
```{r}
write.table(Cluster_2_list[["Cluster1"]],"./RNAseq/HL60/Cluster1_down_genes.txt", row.names = F, col.names = F,quote = F, sep = "\t")

write.table(Cluster_2_list[["Cluster2"]],"./RNAseq/HL60/Cluster2_Up_genes.txt", row.names = F, col.names = F,quote = F, sep = "\t")
```

#### save the genes of the two clusters with TPM and Var

log2(TPM) ---> var. so the var will not be that big

the goal of this part  is hope to help with find potential genes

```{r}
Cluster1_down_genes <- read.csv("./RNAseq/HL60/Cluster1_down_genes.txt",header = F)$V1
length(Cluster1_down_genes)

Cluster2_Up_genes<- read.csv("./RNAseq/HL60/Cluster2_Up_genes.txt",,header = F)$V1
length(Cluster2_Up_genes)

TPM_all <- read.csv("./RNAseq/HL60/TPM.csv")

head(TPM_all)
dim(TPM_all)
TPM_all <- select(TPM_all,-X)
```


```{r}
Cluster1_down_genes_TPM <- TPM_all[which(TPM_all$Gene %in% Cluster1_down_genes ),]
dim(Cluster1_down_genes_TPM)
Cluster1_down_genes_TPM[["log2TPM_var"]] <- apply(log2(Cluster1_down_genes_TPM[,2:dim(Cluster1_down_genes_TPM )[2]]),1,var )

Cluster1_down_genes_TPM<- Cluster1_down_genes_TPM[order(Cluster1_down_genes_TPM$log2TPM_var,decreasing = T),]

head(Cluster1_down_genes_TPM)

write.csv(Cluster1_down_genes_TPM,"./RNAseq/HL60/Cluster1_down_genes_TPM.csv", row.names = F, col.names = T)
```


```{r}
Cluster2_Up_genes_TPM <- TPM_all[which(TPM_all$Gene %in% Cluster2_Up_genes ),]
dim(Cluster2_Up_genes_TPM)

Cluster2_Up_genes_TPM[["log2TPM_var"]] <- apply(log2(Cluster2_Up_genes_TPM[,2:dim(Cluster2_Up_genes_TPM )[2]]),1,var )

Cluster2_Up_genes_TPM<- Cluster2_Up_genes_TPM[order(Cluster2_Up_genes_TPM$log2TPM_var,decreasing = T),]

head(Cluster2_Up_genes_TPM)

write.csv(Cluster2_Up_genes_TPM,"./RNAseq/HL60/Cluster2_Up_genes_TPM.csv", row.names = F, col.names = T)
```

