---
title: "2_Fig2_FigS3_Fig5a_S7_Gene_expression_boxplot_heatmap"
output: 
  html_document:
    number_sections: true
    toc: true
    toc_depth: 3
    toc_float: 
      collapsed: false
      smooth_scroll: false
---


Goal:

look at individual gene expression/gene set pattern among the samples



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "./RNAseq/HL60/")

```

# load library
```{r}
library(dplyr) # use select
library(tidyr)
library(ggplot2)
library(Hmisc)# in order to use capitalize
library('biomaRt')
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
library(purrr)
library("pheatmap")

library(ggpubr) # to use stat_compare_means()

```

```{r}
	plot_theme <-  theme_bw() +
		  theme(axis.line = element_line(colour = "black"),
		        panel.grid.major = element_blank(),
		        panel.grid.minor = element_blank(),
		        panel.border = element_blank(),
		        panel.background = element_blank(),
		        # title in the middle
		        plot.title = element_text(hjust = 0.5),
		        text = element_text(size = 15),axis.text = element_text(size = 15)) +theme(aspect.ratio = 1)

  box_theme <- theme_bw() +
  theme(axis.line = element_line(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.border = element_blank(),
        panel.background = element_blank(),
        # title in the middle
        plot.title = element_text(hjust = 0.5)) 

```

# input the data 
```{r}
## DEG data 

df <- read.csv("./RNAseq/HL60/TPM.csv")

df <- dplyr::select(df,-X)
head(df)

df<-df[,-which(colnames(df)=="RA4h_1_TPM")] 

# just remove the _TPM to make the name shorter !!!
colnames(df)[2:dim(df)[2]] <- unlist(lapply(X = colnames(df)[2:dim(df)[2]],FUN = function(x) {return (strsplit(x,"_TPM")[[1]][1])}))

head(df)
```

```{r}

Cluster_Up<- read.csv("./RNAseq/HL60/Cluster2_Up_genes.txt",header = F)$V1
length(Cluster_Up)

Cluster_Down<- read.csv("./RNAseq/HL60/Cluster1_down_genes.txt",header = F)$V1
length(Cluster_Down)
```


# part 1:  manipulate the data for individual gene plot
```{r}
df_longer <-  df %>% pivot_longer(cols = colnames(df)[-1], names_to = c("Condition","Replicates"),names_pattern= "(.*)_([1-9])", values_to= "TPM" ) %>% as.data.frame()
head(df_longer)
```


### TFs

```{r}

TFs<- c("Gata2","Irf8","Gfi1","Cebpa","Lyl1","Plagl2","Ikzf1","Cebpe","Spi1","Cebpb","Nfil3","Max","Nfe2","Klf5","Mxd1","Ltf","Klf7","Mlx","Xbp1","Mafg","E2f2","Rad21","Cebpz","Myc","Bclaf1","Ets1","Klf13","Zbtb7b","Fosl1","ZNF595","Etv5","Tcf7","E2f1","Nelfe","Pou2af1","Rb1","Cux1","Ezh2","Ybx1","Nfyb","E2f8","Tp53","Nfil3","Klf2","Tbx6","Irf9","Irf2","Irf5","Bcl3","Stat3","Irf7","Stat1","Nfkb2","Atf3","Fosl2","Ddit3","Klf3","Junb","Jdp2","Irf8","Nfe2l2","Irf1","Stat2","Jund","Fosb","Pgr","EGR1","EGR2","EGR3","ID1","FOS","Bcl2","Bcl6","Tbx21","Bcl2l11", "STAT6", "MAFB","REL","MYB")

TFs <-unique(unlist(lapply(X =TFs ,FUN= function(x) {return(toupper(x))})))
TFs
```

#### check 4h DEGs in TFs

```{r}
DEG_4h_up <- read.csv("./RNAseq/HL60/Filter_DEG/Ctrl-less-than-RA4h.csv")$Gene
length(DEG_4h_up)

DEG_4h_down <- read.csv("./RNAseq/HL60/Filter_DEG/Ctrl-more-than-RA4h.csv")$Gene
length(DEG_4h_down)


cat("4h up TFs are: ")
DEG_4h_up[DEG_4h_up%in% TFs] ## "EGR1" "FOS"  "ID1" 
cat("\n")
cat("4h Down TFs are: ")
DEG_4h_down[DEG_4h_down%in% TFs] ## IRF5


```

```{r}
for (gene in TFs){
  # gene <- "NFYC"
  p <- ggplot(df_longer[which(df_longer$Gene== gene),], aes(x = Condition, y = TPM  )) + stat_boxplot(geom ='errorbar', width = 0.4) +geom_boxplot(width = 0.6,outlier.shape = NA)+ geom_jitter(position=position_jitter(0.2))+
    labs(title =gene)+ plot_theme + scale_x_discrete(limits = sample_order)+
  # + ylim(0,max(df_longer[which(df_longer$Gene== gene),][["TPM"]]))+
    theme(text = element_text(size = 20)) + theme(axis.text = element_text(size = 20))  
  
  print(p)
}
```


```{r}
interested_genes2 <- c("GATA1","RARA","RXRA" ,"RUNX1","RUNX2","RUNX3","CDKN1A","NOTCH1","TFE3","ZBTB16")

for (gene in interested_genes2){
  p <- ggplot(df_longer[which(df_longer$Gene== gene),], aes(x = Condition, y = TPM  )) + stat_boxplot(geom ='errorbar', width = 0.4) +geom_boxplot(width = 0.6,outlier.shape = NA)+ geom_jitter(position=position_jitter(0.2))+
    labs(title =gene)+ plot_theme + scale_x_discrete(limits = sample_order)+
  # + ylim(0,max(df_longer[which(df_longer$Gene== gene),][["TPM"]]))+
    theme(text = element_text(size = 10)) + theme(axis.text = element_text(size = 10))  
  
  print(p)
}
```


## paper_related_genes

```{r}

paper_related_genes <- c("ALOX5AP","CYBB","LOXL3","CCL2","RARA","RXRA","ID1","FOS","EGR1","IRF5","CEBPB")

# Groups<- c("Ctrl","RA4h","RA2d","RA4d")

my_comparisons <- rev(list(c("Ctrl","RA4h"),c("RA4h","RA2d"),c("RA2d","RA4d")))

pdf("./paper_related_genes.pdf")
for (gene in paper_related_genes ){
  # gene <- "ALOX5AP"
  p <- ggplot(df_longer[which(df_longer$Gene== gene),], aes(x = Condition, y = TPM  )) + stat_boxplot(geom ='errorbar', width = 0.4) +geom_boxplot(width = 0.6,outlier.shape = NA)+ geom_jitter(position=position_jitter(0.2))+
    labs(title = paste(gene,"\t", "t.test on TPM"))+ plot_theme + scale_x_discrete(limits = sample_order)+
    ylim(0,max(df_longer[which(df_longer$Gene== gene),][["TPM"]])*1.3)+
    stat_compare_means(comparison=my_comparisons,label="p.format",method="t.test")+
    theme(text = element_text(size = 20)) + theme(axis.text = element_text(size = 20))  

  print(p)
  
  
  # anova pvalue 
  anova_test <- anova(lm(log2(TPM+0.1)~Condition, data = df_longer[which(df_longer$Gene== gene),])) 
  anova_test$`Pr(>F)`[1]
  
    p <- ggplot(df_longer[which(df_longer$Gene== gene),], aes(x = Condition, y = TPM  )) + stat_boxplot(geom ='errorbar', width = 0.4) +geom_boxplot(width = 0.6,outlier.shape = NA)+ geom_jitter(position=position_jitter(0.2))+
    labs(title = paste(gene,"\n", "anova test on log2(TPM)","\n",signif(anova_test$`Pr(>F)`[1],3)))+ plot_theme + scale_x_discrete(limits = sample_order)+
    ylim(0,max(df_longer[which(df_longer$Gene== gene),][["TPM"]])*1.3)+
    # stat_compare_means(comparison=my_comparisons,label="p.format",method="t.test")+
    theme(text = element_text(size = 20)) + theme(axis.text = element_text(size = 20))  
  
  
  print(p)
  
}

dev.off()
```


# --------------------------------------------------------------------------------------------

# part 2:  heatmap of gene set (Neu related term or gene groups)


### TFs


```{r}

TFs<- c("Gata2","Irf8","Gfi1","Cebpa","Lyl1","Plagl2","Ikzf1","Cebpe","Spi1","Cebpb","Nfil3","Max","Nfe2","Klf5","Mxd1","Ltf","Klf7","Mlx","Xbp1","Mafg","E2f2","Rad21","Cebpz","Myc","Ets1","Klf13","Zbtb7b","Fosl1","ZNF595","Etv5","Tcf7","E2f1","Nelfe","Pou2af1","Rb1","Cux1","Ezh2","Ybx1","Nfyb","E2f8","Bclaf1","Tp53","Nfil3","Klf2","Tbx6","Irf9","Irf2","Irf5","Bcl3","Stat3","Irf7","Stat1","Nfkb2","Atf3","Fosl2","Ddit3","Klf3","Junb","Jdp2","Irf8","Nfe2l2","Irf1","Stat2","Jund","Fosb","Pgr","EGR1","EGR2","EGR3","ID1","FOS","Bcl2","Bcl6","Tbx21","Bcl2l11","STAT6", "MAFB","REL","MYB")

TFs_Neu2<-unique(unlist(lapply(X =TFs ,FUN= function(x) {return(toupper(x))})))
TFs_Neu2

```


### neutrophil granule-related genes 
------ the expression of neutrophil granule-related genes for all neutrophils
        Gene classes
        Azurophil: c("Mpo", "Elane", "Ctsg", "Prtn3","Prss57", "Ctsc")
        Specific: c("Camp","Ltf","Cybb", "Cyba","Lcn2","B2m")
        Geletinase: c("Mmp8","Mmp9","Hp","Slpi","Itgam")
        Secretory: c("Cd63","Fcgr3","Cd177","Mmp25","Fpr1","Scamp1","Vamp2","Stxbp4")

```{r}
granule_Azurophil <- c("Mpo", "Elane", "Ctsg", "Prtn3","Prss57", "Ctsc")
granule_Azurophil<- unique(unlist(lapply(X =granule_Azurophil ,FUN= function(x) {return(toupper(x))})))
granule_Azurophil

granule_Specific<- c("Camp","Ltf","Cybb", "Cyba","Lcn2","B2m")
granule_Specific<- unique(unlist(lapply(X =granule_Specific ,FUN= function(x) {return(toupper(x))})))
granule_Specific

granule_Geletinase<- c("Mmp8","Mmp9","Hp","Slpi","Itgam")
granule_Geletinase <-unique(unlist(lapply(X =granule_Geletinase ,FUN= function(x) {return(toupper(x))})))
granule_Geletinase

granule_Secretory<- c("Cd63","Cd177","Mmp25","Fpr1","Scamp1","Vamp2","Stxbp4")
granule_Secretory<- unique(unlist(lapply(X =granule_Secretory ,FUN= function(x) {return(toupper(x))})))
granule_Secretory
```


### Aging related genes


```{r}
Aging <- c("ICAM1","ITGA4","CD47","ITGAM","CXCR2","SELL","CXCR4","TLR4","ITGAX")
```

### cell cycle related genes

```{r}
cell_cycle <- c("CKS1","MKI67","RANBP1","SPC24","ANP32B","H2AFX","CCNE2","CDC20","SMC2","PMF1","TOP2A","ESCO2",
                "CKS1b","CENPF","CDKN3","TUBA1C","KIF22","STMN1","SMC4")
```


### get interested GO term genes
----- phagocytosis, engulfmentâ€™ (GO: 0006911)
      chemotaxis score (GO:0030593), 
      neutrophil activation score (GO:0042119)
      NADPH oxidase score for each cluster. ?
      Heatmap showing relative expression of **seven genes of the NADPH oxidase** complex for all neutrophils
      c("Cybb","Cyba","Rac2","Rac1","Ncf2","Ncf1","Ncf4")
      
      I checked the term NADPH oxidase complex,the GO is  **GO:0043020**
      
```{r}
phagocytosis <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0006911',
      mart = human)

phagocytosis<- phagocytosis$hgnc_symbol
phagocytosis <- unique(phagocytosis[phagocytosis != ""])
length(phagocytosis)
print(phagocytosis)

cat(phagocytosis,sep = '\",\"' )
```

```{r}
chemotaxis <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0030593',
      mart = human)

chemotaxis<- chemotaxis$hgnc_symbol
chemotaxis <- unique(chemotaxis[chemotaxis != ""])
length(chemotaxis)
print(chemotaxis)

cat(chemotaxis,sep = '\",\"' )
```

```{r}
neutrophil_activation  <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0042119',
      mart = human)

neutrophil_activation<- neutrophil_activation$hgnc_symbol
neutrophil_activation <- unique(neutrophil_activation[neutrophil_activation != ""])
length(neutrophil_activation)
print(neutrophil_activation)
cat(neutrophil_activation,sep = '\",\"' )
```

```{r}
NADPH_oxidase <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0043020',
      mart = human)

NADPH_oxidase<- NADPH_oxidase$hgnc_symbol
NADPH_oxidase <- unique(NADPH_oxidase[NADPH_oxidase != ""])
length(NADPH_oxidase)
print(NADPH_oxidase)

cat(NADPH_oxidase,sep = '\",\"' )
```
      
  
   ------metabolism:
      Glycolysis (Reactome Pathway Database #R-MMU-70171); 
            canonical glycolysis: GO:0061621 (less genes )
            glycolytic process: GO:0006096 (more genes )
       Oxidative phosphorylation (GO:000619);  aa typo: Oxidative_phosphorylation GO:0006119
       Electron transport chain (GO:0022900); 
       Tricarboxylic acid cycle (GO:0006099).

```{r}
canonical_glycolysis <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0061621',
      mart = human)

canonical_glycolysis<- canonical_glycolysis$hgnc_symbol
canonical_glycolysis <- unique(canonical_glycolysis[canonical_glycolysis != ""])
length(canonical_glycolysis)
print(canonical_glycolysis)

cat(canonical_glycolysis,sep = '\",\"' )
```
    
    ```{r}
glycolytic_process <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0006096',
      mart = human)

glycolytic_process<- glycolytic_process$hgnc_symbol
glycolytic_process <- unique(glycolytic_process[glycolytic_process != ""])
length(glycolytic_process)
print(glycolytic_process)

cat(glycolytic_process,sep = '\",\"' )
```
    
    ```{r}
Oxidative_phosphorylation <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0006119',
      mart = human)

Oxidative_phosphorylation<- Oxidative_phosphorylation$hgnc_symbol
Oxidative_phosphorylation <- unique(Oxidative_phosphorylation[Oxidative_phosphorylation != ""])
length(Oxidative_phosphorylation)
print(Oxidative_phosphorylation)

cat(Oxidative_phosphorylation,sep = '\",\"' )
```
    
    ```{r}
Electron_transport_chain <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0022900',
      mart = human)

Electron_transport_chain<- Electron_transport_chain$hgnc_symbol
Electron_transport_chain <- unique(Electron_transport_chain[Electron_transport_chain != ""])
length(Electron_transport_chain)
print(Electron_transport_chain)

cat(Electron_transport_chain,sep = '\",\"' )
```
    
    ```{r}
Tricarboxylic_acid_cycle <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0006099',
      mart = human)

Tricarboxylic_acid_cycle<- Tricarboxylic_acid_cycle$hgnc_symbol
Tricarboxylic_acid_cycle <- unique(Tricarboxylic_acid_cycle[Tricarboxylic_acid_cycle != ""])
length(Tricarboxylic_acid_cycle)
print(Tricarboxylic_acid_cycle)

cat(Tricarboxylic_acid_cycle,sep = '\",\"' )
```
### granulocyte migration GO:0097530  


```{r}
GranMig <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:0097530',
      mart = human)

GranMig<- GranMig$hgnc_symbol
GranMig <- unique(GranMig[GranMig != ""])
length(GranMig)
print(GranMig)

cat(GranMig,sep = '\",\"' )
```

neutrophil migration: 1990266


```{r}
NeuMig <- getBM(attributes = c('entrezgene_id','hgnc_symbol'),
      filters = 'go',
      values = 'GO:1990266',
      mart = human)

NeuMig<- NeuMig$hgnc_symbol
NeuMig <- unique(NeuMig[NeuMig != ""])
length(NeuMig)
print(NeuMig)

cat(NeuMig,sep = '\",\"' )
```


## plot heatmap of interested genes

### define function
```{r}
gene_set_heatmap <- function(GeneSetName){
  # GeneSetName <- "TFs_Neu"
  GeneSet<- get(GeneSetName)
  df_temp <- df[which(df$Gene %in% GeneSet),]
  rownames(df_temp)<- df_temp$Gene
  df_temp<-df_temp[GeneSet,]
  df_temp <- df_temp[which(apply(df_temp[,c(-1)],1,var) != 0),]
 	mtx <- as.matrix(df_temp[,c(-1)])
  rownames(mtx) <- df_temp$Gene
  mtx_colnames <-  colnames(df_temp[,c(-1)])
  fun1 <- function(x) {return (substr(x,1,gregexpr(pattern ='_',x)[[1]][1]-1))}
  Conditions <- (unlist(lapply(mtx_colnames,fun1)))

  Sample_info <- data.frame(Conditions  = factor(Conditions,levels = c("Ctrl","RA4h","RA2d","RA4d")))

  ##  IMPORTANT ******* here must be consistent ********!!!!!!!

  rownames(Sample_info) <- mtx_colnames
  Sample_info
  dim(mtx)
  #  hierarchy- cluster
  pheatmap(mtx ,cluster_rows=T,cluster_cols = F,
           # show the gene name
           show_rownames = T,
           scale = "row",
           # cutree_rows=5,
           angle_col='90',
           # annotation_row = gene_anno ,
           annotation_col = Sample_info,
           fontsize_col=10,
           main = paste("hierarchy cluster \n",GeneSetName))
  
  mtx_Zscore <- (mtx-apply(X = mtx, MARGIN = 1, mean))/apply(X = mtx, MARGIN = 1, sd)
  is.na(mtx_Zscore) %>% table()

  set.seed(2);mtx_kmeans <- kmeans(mtx_Zscore,centers=2,nstart = 25)

  gene_kmeans <- as.data.frame(sort(mtx_kmeans$cluster)); names(gene_kmeans) <- 'cluster';
  gene_kmeans$cluster <- factor(gene_kmeans$cluster)

  pheatmap(mtx_Zscore[names(sort(mtx_kmeans$cluster)),],cluster_rows=F,cluster_cols = F,angle_col='315',annotation_row = gene_kmeans, annotation_col = Sample_info,main = 'z-score (k-means)',show_rownames = T,fontsize_col=10,fontsize_row = 8,)

  mtx_Zscore<- as.data.frame(mtx_Zscore)
  
  mtx_Zscore[["Gene"]] <- rownames(mtx_Zscore)
  
  mtx_Zscore_long <- mtx_Zscore %>% pivot_longer(cols = colnames(mtx_Zscore)[1:(length(colnames(mtx_Zscore)) -1)], names_to = "Samples",values_to= "TPM_Zscore" ) %>% as.data.frame()
  
  fun1<- function(x) {return  (unlist(strsplit(x, "[_]"))[1])}
  mtx_Zscore_long$Samples<- unlist(lapply(mtx_Zscore_long$Samples,fun1))
  mtx_Zscore_long$Samples <-factor(Conditions,levels = c("Ctrl","RA4h","RA2d","RA4d"))
  
  head(mtx_Zscore_long)
  
  Median <- mtx_Zscore_long %>% dplyr::group_by(Samples) %>% dplyr::summarize(Median = median(TPM_Zscore))
  
  p <- ggplot(data =mtx_Zscore_long,aes(x = Samples,y = TPM_Zscore) )+ geom_boxplot() +labs(title = paste("Zscore \n",GeneSetName))+
      geom_point(data = Median, mapping = aes(x = Samples, y = Median)) +
      # , group=1 is necessary
    geom_line(data = Median, mapping = aes(x = Samples, y = Median, group=1),col = "red") + plot_theme
    
  print(p) 
  
}
```


## plot all the terms
```{r}
granules <- c(granule_Azurophil,granule_Specific,granule_Geletinase,granule_Secretory)
intrested_gene_sets <- c("NL","TFs_Neu","TFs_Neu2","granules","phagocytosis","chemotaxis","neutrophil_activation","NADPH_oxidase","canonical_glycolysis","glycolytic_process","Oxidative_phosphorylation","Electron_transport_chain","Tricarboxylic_acid_cycle","Aging","cell_cycle","NeuMig")
```


```{r}
for (GeneSetName in intrested_gene_sets){
  gene_set_heatmap(GeneSetName)
}

```


## plot TF 


```{r}
  GeneSetName <- "TFs_Neu2"
  GeneSet<- get(GeneSetName)
  length(GeneSet)
  df_temp <- df[which(df$Gene %in% GeneSet),]
  dim(df_temp)
  GeneSet[which(GeneSet %in% df_temp$Gene== FALSE)]
  
  # order the data beased on the  gene in the GeneSet
  rownames(df_temp)<- df_temp$Gene
  
  df_temp<-df_temp[GeneSet,]

  df_temp <- df_temp[which(apply(df_temp[,c(-1)],1,var) != 0),]
 
	mtx <- as.matrix(df_temp[,c(-1)])
  rownames(mtx) <- df_temp$Gene
  head(mtx)
  
  mtx_colnames <-  colnames(df_temp[,c(-1)])
  mtx_colnames

  fun1 <- function(x) {return (substr(x,1,gregexpr(pattern ='_',x)[[1]][1]-1))}
  Conditions <- (unlist(lapply(mtx_colnames,fun1)))

  Sample_info <- data.frame(Conditions  = factor(Conditions,levels = c("Ctrl","RA4h","RA2d","RA4d")))


  rownames(Sample_info) <- mtx_colnames

  Sample_info
  
 dim(mtx)

  
  pdf("./TFs_heatmap.pdf",width = 5,height = 15)
  
  pheatmap(mtx ,cluster_rows=T,cluster_cols = F,
           # show the gene name
           show_rownames = T,
           scale = "row",
           # cutree_rows=5,
           angle_col='90',
           # annotation_row = gene_anno ,
           annotation_col = Sample_info,
           fontsize_col=10,fontsize_row = 8,
           main = paste("hierarchy cluster \n",GeneSetName))
    dev.off()

  # ------- kmeans--------- 
  pdf("./TFs_heatmap_kmeans.pdf",width = 5,height = 15) 
  mtx_Zscore <- (mtx-apply(X = mtx, MARGIN = 1, mean))/apply(X = mtx, MARGIN = 1, sd)
  dim(mtx_Zscore)
  set.seed(2);mtx_kmeans <- kmeans(mtx_Zscore,centers=6,nstart = 25)
  
  gene_kmeans <- as.data.frame(sort(mtx_kmeans$cluster)); names(gene_kmeans) <- 'cluster'; 
  gene_kmeans$cluster <- factor(gene_kmeans$cluster)
    
  pheatmap(mtx_Zscore[names(sort(mtx_kmeans$cluster)),],cluster_rows=F,cluster_cols = F,angle_col='315',annotation_row = gene_kmeans, annotation_col = Sample_info,main = 'z-score (k-means)',
           show_rownames = T,fontsize_col=10,fontsize_row = 8,)
  dev.off()
```

